### PR TITLE
remove `threadedregion` and move `jl_threading_run` to julia

### LINKED ISSUE #32970: Threads are disabled if an error occurs
When an error is thrown in a threaded loop, only one thread is used in threaded loops thereafter.
```julia
julia> function f!(a)
           Threads.@threads for i in eachindex(a)
               a[i] = Threads.threadid()
           end
           a
       end
f! (generic function with 1 method)

julia> function g!(a)
           Threads.@threads for i in eachindex(a)
               sleep(Threads.threadid())
               a[i] = log(-Threads.threadid())
           end
           a
       end
g! (generic function with 1 method)

julia> a = zeros(Threads.nthreads())
6-element Array{Float64,1}:
 0.0
 0.0
 0.0
 0.0
 0.0
 0.0

julia> f!(a)
6-element Array{Float64,1}:
 1.0
 2.0
 3.0
 4.0
 5.0
 6.0

julia> g!(a)
ERROR: TaskFailedException:
DomainError with -1.0:
log will only return a complex result if called with a complex argument. Try log(Complex(x)).
Stacktrace:
 [1] throw_complex_domainerror(::Symbol, ::Float64) at ./math.jl:32
 [2] log(::Float64) at ./special/log.jl:285
 [3] macro expansion at ./special/log.jl:395 [inlined]
 [4] (::var"##19#threadsfor_fun#4"{Array{Float64,1},Base.OneTo{Int64}})(::Bool) at ./threadingconstructs.jl:64
 [5] (::var"##19#threadsfor_fun#4"{Array{Float64,1},Base.OneTo{Int64}})() at ./threadingconstructs.jl:31
Stacktrace:
 [1] wait(::Task) at ./task.jl:251
 [2] macro expansion at ./threadingconstructs.jl:75 [inlined]
 [3] g!(::Array{Float64,1}) at ./REPL[2]:2
 [4] top-level scope at REPL[5]:1

julia> a
6-element Array{Float64,1}:
 1.0
 2.0
 3.0
 4.0
 5.0
 6.0

julia> f!(a)
6-element Array{Float64,1}:
 1.0
 1.0
 1.0
 1.0
 1.0
 1.0

julia> Threads.nthreads()
6
```
This is on Julia v1.3.0-rc1.0. Julia v1.2.0-rc2.0 behaves similar (only a slightly different error message). I'm using the generic Linux x86_64 binaries.

As mentioned by @c42f, this is a bug.

### PR DESCRIPTION
**Pull request by JeffBezanson**
Putting this up to see if we can find any adversarial test cases.

Resolves #32970

### CODE REVIEW

**Comment by staticfloat** [comment]
Wow, this is great!  Using the benchmarks from #32181, we get:

```
julia-master with 1 threads:
#= /Users/sabae/src/julia/speedtest.jl:22 =# @benchmark(func(2.0, 1 << 10)) = Trial(24.241 μs)
#= /Users/sabae/src/julia/speedtest.jl:23 =# @benchmark(func_threaded(2.0, 1 << 10)) = Trial(28.450 μs)
julia-master with 2 threads:
#= /Users/sabae/src/julia/speedtest.jl:22 =# @benchmark(func(2.0, 1 << 10)) = Trial(25.977 μs)
#= /Users/sabae/src/julia/speedtest.jl:23 =# @benchmark(func_threaded(2.0, 1 << 10)) = Trial(17.003 μs)
```

Versus:
```
julia-threadedregion with 1 threads:
#= /Users/sabae/src/julia/speedtest.jl:22 =# @benchmark(func(2.0, 1 << 10)) = Trial(24.241 μs)
#= /Users/sabae/src/julia/speedtest.jl:23 =# @benchmark(func_threaded(2.0, 1 << 10)) = Trial(25.486 μs)
julia-threadedregion with 2 threads:
#= /Users/sabae/src/julia/speedtest.jl:22 =# @benchmark(func(2.0, 1 << 10)) = Trial(24.250 μs)
#= /Users/sabae/src/julia/speedtest.jl:23 =# @benchmark(func_threaded(2.0, 1 << 10)) = Trial(13.774 μs)
```

Not only did you significantly shrink the overhead amount, you sped up the multithreaded case!

**Comment by staticfloat** [comment]
The testing failures were all timeouts so I've restarted them to try and get a feel for how reliable those timeouts are.  I've found some threading lock-ups in my explorations as well, opened an issue here: https://github.com/JuliaLang/julia/issues/32511

File: base/threadingconstructs.jl [hunk 1]
```diff
@@ -65,20 +62,20 @@ function _threadsfor(iter,lbody)
...
-        if threadid() != 1 || in_threaded_loop[]
-            # We are in a nested threaded loop
-            Base.invokelatest(threadsfor_fun, true)
-        else
-            in_threaded_loop[] = true
-            # the ccall is not expected to throw
-            ccall(:jl_threading_run, Cvoid, (Any,), threadsfor_fun)
-            in_threaded_loop[] = false
-        end
+        threading_run(threadsfor_fun)
```
Comment (staticfloat) [inline base/threadingconstructs.jl]: Because we are never invoking `threadsfor_fun` with an argument, should the `onethread` argument be removed?

**Comment by vtjnash** [comment]
> Wow, this is great!

One possible cause for that is that this additionally deletes the IO-capable scheduler and always instead uses the multi-threading-optimized scheduler. We currently may have a small lag when switching, while this just instead adds a large lag (latency) for any actual IO work.

**Comment by JeffBezanson** [comment]
> while this just instead adds a large lag (latency) for any actual IO work.

Right; part of the goal of this PR was to motivate a benchmark for that and then try to address it. So, do your worst! :)

**Comment by JeffBezanson** [comment]
Wait, we are still polling `jl_process_events` in the task_get_next loop, so I don't think there is any extra I/O latency (just slightly slower task switching).

**Comment by oschulz** [comment]
Earlier, I saw some code ```ccall(:jl_set_task_tid, ...)``` in this PR. Would it maybe be useful to expose an official method like

```
function schedule(t::Task, tid)
    @assert !istaskstarted(t)
    t.sticky = true
    ccall(:jl_set_task_tid, Cvoid, (Any, Cint), t, tid-1)
    schedule(t)
    return t
end
```

to the user, for applications that require more direct control over what's run on which thread?

**Comment by oschulz** [comment]
A possible application for such a `schedule(t::Task, tid)` would be benchmarks across NUMA domains (with thread-pinning), etc.

**Comment by JeffBezanson** [comment]
Yes, that's possible to add. My thinking was just that it's not a good default for `@threads`, since it's not as composable.

**Comment by oschulz** [comment]
> My thinking was just that it's not a good default for @threads, since it's not as composable.

Oh sure - I was really thinking about specialized use cases. Certainly not something the average user will require, but HPC applications may sometimes need it, e.g. for benchmarking and performance analysis, etc. For such cases, I think it would be helpful to have an "offical" API, like `schedule(t::Task, tid)` - assuming `:jl_set_task_tid` or an equivalent will continue to exist.

**Comment by mbauman** [comment]
I'd really love to see this merged — or otherwise make the "default/easy/obvious" `for` loop macro do this in some way.  As noted by Jacob in the competing PR, it's unfortunately breaking in the wild.

https://github.com/JuliaLang/julia/pull/35003#issuecomment-618132742

> I think it's been mentioned elsewhere, but just wanted to add a voice that this would indeed be breaking; a few of us _rely_ on the current behavior in order to start worker tasks on specific threads, such as the code [here](https://github.com/tro3/ThreadPools.jl/blob/e36e82c51adfd22a7d0459912861bfaf1d43deff/src/qpool.jl#L16). I'd say as long as we have either: 1) a way to reserve a main thread where tasks couldn't be spawned, or 2) a way to still spawn a threaded task on a specific thread (e.g. `Threads.@spawn 2 expr`), then we could still get the behavior we're after.

**Comment by JeffBezanson** [comment]
Is the appealing part the dynamic schedule? We could get that in a less invasive way just by flipping the sticky bit within the current implementation.

**Comment by mbauman** [comment]
The main thing is that I want to be able to throw `@threads for` inside of libraries and not worry about oversubscribing if an app also uses `@threads for`.

**Comment by JeffBezanson** [comment]
There won't be any oversubscription either way (except via a native library), but I assume the real problem is the schedule --- we don't want `@threads` to pin tasks to specific threads by default?

**Comment by mbauman** [comment]
Right, of course, I wasn't thinking straight.  Yes, the scheduling won't be as ideal, but I also have it in my head that nested `@threads` loops are crashy and not the best idea.  Is that better these days?

**Comment by JeffBezanson** [comment]
Should not be crashy.

**Comment by OkonSamuel** [comment]
@JeffBezanson. Just wanted to clarify. I used to be of the opinion that nesting `@threads for` calls would oversubcribe CPU. Are you saying that apart from the dynamic scheduling issue `@threads for` won't oversubcribe ?

**Comment by JeffBezanson** [comment]
Correct, `@threads` will never oversubscribe your CPUs. The number of threads we use is fixed at startup, and `@threads` and `@spawn` just create tasks to run on those threads.

**Comment by ranocha** [comment]
Are there any plans when this will be merged? It would be really nice to get the speed-up when using `Threads.@threads for ...` using a single thread, cf. https://discourse.julialang.org/t/overhead-of-threads-threads/53964.

**Comment by vtjnash** [comment]
This was moved to Julia (and so the PR is conflicted)

**Pull request closed without merging**